crates/p2p/src/main_loop.rs

```diff
@@ -207,11 +258,22 @@ impl MainLoop {
                 peer_id,
                 num_established,
                 connection_id: _, // TODO consider tracking connection IDs for peers
+                endpoint,
                 ..
             } => {
                 if num_established == 0 {
                     self.peers.write().await.peer_disconnected(&peer_id);
+                    // Don't keep expired peers in the recent peers set.
+                    if let Some(peer_ip) = get_ip(endpoint.get_remote_address()) {
+                        self.recent_direct_peers.remove_if_expired(&peer_ip);
```

Comment:
Ok guys, so if I understood your discussion correctly this call should:
- remove the entry immediately if it already expired 
- "start a timer" (whatever that means implementation wise) for the entry to be removed when expiry deadline is hit